src/service/position-observer-impl.js

```diff
@@ -0,0 +1,413 @@
+          entry.handler(position);
+        }
+      },
+    };
+
+    this.entries_.push(entry);
+
+    if (this.entries_.length == 1) {
+      this.startCallback();
+    }
+
+    this.vsync_.measure(() => {
+      this.updateEntryPosition_(entry);
+    });
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (entry.element === element) {
+        entry.fidelity = fidelity;
+        if (fidelity == PositionObserverFidelity.HIGH) {
+          entry.turn = 0;
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot change fidelity on unobserved element');
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  unobserve(element) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      if (this.entries_[i].element === element) {
+        this.entries_[i].handler = null;
+        this.entries_.splice(i, 1);
+        if (this.entries_.length == 0) {
+          this.stopCallback();
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot unobserve unobserved element');
+  }
+
+  /** @param {!Object} unusedEntry */
+  updateEntryPosition_(unusedEntry) {}
+
+  startCallback() {}
+
+  stopCallback() {}
+
+  increaseFidelityCallback() {}
+
+  decreaseFidelityCallback() {}
+}
+
+export class AmpDocPositionObserver extends AbstractPositionObserver {
+
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+
+    super(ampdoc);
+
+    /** @private {boolean} */
+    this.inScroll_ = false;
+
+    /** @private {boolean} */
+    this.measure_ = false;
+
+    /** @private {number} */
+    this.scrollTimer_ = Date.now();
+
+    /** @private {Array<function()>} */
+    this.unlisteners_ = [];
+  }
+
+  /* @override */
+  startCallback() {
+    // listen to viewport scroll event to help pass determine if need to
+    this.unlisteners_.push(this.viewport_.onScroll(() => {
+      this.inScroll_ = true;
+      this.scrollTimer_ = Date.now();
+      this.schedulePass_();
+      setTimeout(() => {
+        if (Date.now() - this.scrollTimer_ < 500) {
+          // viewport scroll in the last 500 ms, wait
+          return;
+        }
+        // assume scroll stops, if no scroll event in the last 500,
+        this.inScroll_ = false;
+      }, 500);
+    }));
+    this.unlisteners_.push(this.viewport_.onChanged(() => {
+      this.vsync_.measure(() => {
+        this.pass_(true);
+      });
+    }));
+    // QQQ: Already call this.updateEntryPosition_(entry) in observe()? Needed?
+    // initial pass
+    this.vsync_.measure(() => {
+      this.pass_(true);
+    });
+  }
+
+  /* @override */
+  stopCallback() {
+    while (this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /** @param {boolean=} recursive */
+  schedulePass_(recursive) {
```

Comment:
> is to check if a pass is already scheduled and I should be able to remove recursive

👍

> But I would still need this.inScroll_ to stop after scroll has stopped.

That's where a proper `debounce` will come in. It only calls the callback after `wait` ms have elapsed since the last call.